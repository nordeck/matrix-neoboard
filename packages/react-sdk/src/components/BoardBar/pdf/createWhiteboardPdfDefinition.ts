/*
 * Copyright 2023 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import '@fontsource/noto-emoji/400.css';
import { Content, TDocumentDefinitions } from 'pdfmake/interfaces';
import { isDefined } from '../../../lib';
import { WhiteboardInstance } from '../../../state';
import { whiteboardHeight, whiteboardWidth } from '../../Whiteboard';
import { createWhiteboardPdfContentSlide } from './createWhiteboardPdfContentSlide';
import { forceLoadFontFamily } from './forceLoadFontFamily';

export async function createWhiteboardPdfDefinition({
  whiteboardInstance,
  roomName,
  authorName,
  baseUrl,
}: {
  whiteboardInstance: WhiteboardInstance;
  roomName: string;
  authorName: string;
  baseUrl: string;
}): Promise<TDocumentDefinitions> {
  // make sure the font is loaded so the text size calculations are correct
  await forceLoadFontFamily('Noto Emoji');
  const whiteboardExport = await whiteboardInstance.export(baseUrl);

  return {
    pageMargins: 0,
    pageSize: { width: whiteboardWidth, height: whiteboardHeight },
    content: whiteboardInstance
      .getSlideIds()
      .map<Content | undefined>((slideId, idx) => {
        const slide = whiteboardInstance.getSlide(slideId);

        if (slide) {
          return {
            stack: [
              createWhiteboardPdfContentSlide(
                slide,
                whiteboardExport.whiteboard.files,
              ),
            ],
            pageBreak: idx > 0 ? 'before' : undefined,
          };
        }

        return undefined;
      })
      .filter(isDefined),
    version: '1.5',
    styles: {
      emoji: {
        font: 'Noto Emoji',
      },
    },
    info: {
      title: roomName,
      author: authorName,
    },
  };
}
