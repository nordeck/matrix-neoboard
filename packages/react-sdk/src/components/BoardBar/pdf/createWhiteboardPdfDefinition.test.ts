/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { mockWidgetApi } from '@matrix-widget-toolkit/testing';
import {
  mockEllipseElement,
  mockLineElement,
  mockWhiteboardManager,
} from '../../../lib/testUtils/documentTestUtils';
import { createWhiteboardPdfDefinition } from './createWhiteboardPdfDefinition';
import * as font from './forceLoadFontFamily';

jest.mock('./forceLoadFontFamily', () => {
  return {
    __esModule: true,
    ...jest.requireActual('./forceLoadFontFamily'),
  };
});

describe('createWhiteboardPdfDefinition', () => {
  beforeEach(() => {
    jest
      .spyOn(Element.prototype, 'getBoundingClientRect')
      .mockImplementation(() => ({
        width: 100,
        height: 100,
        x: 0,
        y: 0,
        bottom: 0,
        left: 0,
        right: 0,
        top: 0,
        toJSON: jest.fn(),
      }));
  });

  it('should generate a pdf header and table', async () => {
    const { whiteboardManager } = mockWhiteboardManager({
      slides: [
        [
          'slide-0',
          [
            [
              'element-rectangle',
              mockEllipseElement({ kind: 'rectangle', text: 'Rectangle' }),
            ],
            [
              'element-rectangle-text',
              mockEllipseElement({
                kind: 'rectangle',
                text: 'Rectangle Text',
                fillColor: 'transparent',
              }),
            ],
            [
              'element-circle',
              mockEllipseElement({
                kind: 'circle',
                text: 'Circle ✅',
                textAlignment: 'right',
              }),
            ],
            [
              'element-ellipse',
              mockEllipseElement({ kind: 'ellipse', text: 'Ellipse' }),
            ],
            [
              'element-triangle',
              mockEllipseElement({ kind: 'triangle', text: 'Triangle 👍' }),
            ],
          ],
        ],
        [
          'slide-1',
          [
            ['element-line', mockLineElement({ kind: 'line' })],
            ['element-polyline', mockLineElement({ kind: 'polyline' })],
          ],
        ],
      ],
    });
    const whiteboardInstance = whiteboardManager.getActiveWhiteboardInstance()!;

    expect(
      await createWhiteboardPdfDefinition({
        whiteboardInstance,
        roomName: 'My Room',
        authorName: 'Alice',
        widgetApi: mockWidgetApi(),
      }),
    ).toMatchSnapshot();
  });

  it('should load "Noto Emoji"', async () => {
    const spy = jest.spyOn(font, 'forceLoadFontFamily');

    const { whiteboardManager } = mockWhiteboardManager();
    const whiteboardInstance = whiteboardManager.getActiveWhiteboardInstance()!;

    await createWhiteboardPdfDefinition({
      whiteboardInstance,
      roomName: 'My Room',
      authorName: 'Alice',
      widgetApi: mockWidgetApi(),
    });

    expect(spy).toHaveBeenCalledWith('Noto Emoji');
  });
});
