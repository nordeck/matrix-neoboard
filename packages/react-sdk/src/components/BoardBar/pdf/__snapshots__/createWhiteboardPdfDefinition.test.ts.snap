// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createWhiteboardPdfDefinition > should generate a pdf header and table 1`] = `
{
  "content": [
    {
      "pageBreak": undefined,
      "stack": [
        [
          [
            {
              "absolutePosition": {
                "x": 0,
                "y": 0,
              },
              "canvas": [
                {
                  "color": "#ffffff",
                  "h": 100,
                  "lineColor": "#ffffff",
                  "lineWidth": 2,
                  "r": undefined,
                  "strokeOpacity": 1,
                  "type": "rect",
                  "w": 50,
                  "x": 0,
                  "y": 1,
                },
              ],
              "unbreakable": true,
            },
            {
              "absolutePosition": {
                "x": 10,
                "y": 28.75,
              },
              "layout": "noBorders",
              "table": {
                "body": [
                  [
                    {
                      "alignment": "center",
                      "color": "#000",
                      "font": "Inter",
                      "fontSize": 10,
                      "lineHeight": 1,
                      "margin": 0,
                      "text": [
                        "Rectangle",
                      ],
                    },
                  ],
                ],
                "heights": [
                  62.25,
                ],
                "widths": [
                  30,
                ],
              },
              "unbreakable": true,
            },
          ],
          [
            {
              "absolutePosition": {
                "x": 0,
                "y": 0,
              },
              "canvas": [
                {
                  "color": undefined,
                  "h": 100,
                  "lineColor": undefined,
                  "lineWidth": 2,
                  "r": undefined,
                  "strokeOpacity": 0,
                  "type": "rect",
                  "w": 50,
                  "x": 0,
                  "y": 1,
                },
              ],
              "unbreakable": true,
            },
            {
              "absolutePosition": {
                "x": 10,
                "y": 28.75,
              },
              "layout": "noBorders",
              "table": {
                "body": [
                  [
                    {
                      "alignment": "center",
                      "color": "#000",
                      "font": "Inter",
                      "fontSize": 10,
                      "lineHeight": 1,
                      "margin": 0,
                      "text": [
                        "Rectangle Text",
                      ],
                    },
                  ],
                ],
                "heights": [
                  62.25,
                ],
                "widths": [
                  30,
                ],
              },
              "unbreakable": true,
            },
          ],
          [
            {
              "absolutePosition": {
                "x": 0,
                "y": 0,
              },
              "canvas": [
                {
                  "color": "#ffffff",
                  "lineColor": "#ffffff",
                  "lineWidth": 2,
                  "r1": 25,
                  "r2": 50,
                  "type": "ellipse",
                  "x": 25,
                  "y": 51,
                },
              ],
              "unbreakable": true,
            },
            {
              "absolutePosition": {
                "x": 7.322330470336311,
                "y": 30.458869925047622,
              },
              "layout": "noBorders",
              "table": {
                "body": [
                  [
                    {
                      "alignment": "right",
                      "color": "#000",
                      "font": "Inter",
                      "fontSize": 10,
                      "lineHeight": 1,
                      "margin": 0,
                      "text": [
                        "Circle ",
                        {
                          "style": "emoji",
                          "text": "‚úÖ",
                        },
                      ],
                    },
                  ],
                ],
                "heights": [
                  55.896469134279755,
                ],
                "widths": [
                  35.35533905932738,
                ],
              },
              "unbreakable": true,
            },
          ],
          [
            {
              "absolutePosition": {
                "x": 0,
                "y": 0,
              },
              "canvas": [
                {
                  "color": "#ffffff",
                  "lineColor": "#ffffff",
                  "lineWidth": 2,
                  "r1": 25,
                  "r2": 50,
                  "type": "ellipse",
                  "x": 25,
                  "y": 51,
                },
              ],
              "unbreakable": true,
            },
            {
              "absolutePosition": {
                "x": 7.322330470336311,
                "y": 30.394660940672622,
              },
              "layout": "noBorders",
              "table": {
                "body": [
                  [
                    {
                      "alignment": "center",
                      "color": "#000",
                      "font": "Inter",
                      "fontSize": 13,
                      "lineHeight": 1,
                      "margin": 0,
                      "text": [
                        "Ellipse",
                      ],
                    },
                  ],
                ],
                "heights": [
                  55.960678118654755,
                ],
                "widths": [
                  35.35533905932738,
                ],
              },
              "unbreakable": true,
            },
          ],
          [
            {
              "absolutePosition": {
                "x": 0,
                "y": 0,
              },
              "canvas": [
                {
                  "closePath": true,
                  "color": "#ffffff",
                  "lineColor": "#ffffff",
                  "lineWidth": 2,
                  "points": [
                    {
                      "x": 0,
                      "y": 101,
                    },
                    {
                      "x": 25,
                      "y": 1,
                    },
                    {
                      "x": 50,
                      "y": 101,
                    },
                  ],
                  "type": "polyline",
                },
              ],
              "unbreakable": true,
            },
            {
              "absolutePosition": {
                "x": 8.333333333333332,
                "y": 70.48087565104166,
              },
              "layout": "noBorders",
              "table": {
                "body": [
                  [
                    {
                      "alignment": "center",
                      "color": "#000",
                      "font": "Inter",
                      "fontSize": 10,
                      "lineHeight": 1,
                      "margin": 0,
                      "text": [
                        "Triangle ",
                        {
                          "style": "emoji",
                          "text": "üëç",
                        },
                      ],
                    },
                  ],
                ],
                "heights": [
                  20.519124348958343,
                ],
                "widths": [
                  33.333333333333336,
                ],
              },
              "unbreakable": true,
            },
          ],
        ],
      ],
    },
    {
      "pageBreak": "before",
      "stack": [
        [
          {
            "absolutePosition": {
              "x": 0,
              "y": 0,
            },
            "canvas": [
              {
                "lineColor": "#ffffff",
                "lineWidth": 4,
                "type": "line",
                "x1": 0,
                "x2": 2,
                "y1": 2,
                "y2": 4,
              },
            ],
            "unbreakable": true,
          },
          {
            "absolutePosition": {
              "x": 0,
              "y": 0,
            },
            "canvas": [
              {
                "lineColor": "#ffffff",
                "lineWidth": 4,
                "points": [
                  {
                    "x": 0,
                    "y": 2,
                  },
                  {
                    "x": 2,
                    "y": 4,
                  },
                ],
                "type": "polyline",
              },
            ],
            "unbreakable": true,
          },
        ],
      ],
    },
  ],
  "info": {
    "author": "Alice",
    "title": "My Room",
  },
  "pageMargins": 0,
  "pageSize": {
    "height": 1080,
    "width": 1920,
  },
  "styles": {
    "emoji": {
      "font": "Noto Emoji",
    },
  },
  "version": "1.5",
}
`;
